analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -457,6 +476,51 @@ public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
             ValidateSetBoolConstraint(isPattern, OperationKindEx.BinaryPattern, expectedBoolConstraint);
 
         private static void ValidateSetBoolConstraint(string isPattern, OperationKind expectedOperation, bool? expectedBoolConstraint)
+        {
+            var validator = CreateSetBoolConstraintValidator(isPattern);
+            validator.ValidateContainsOperation(expectedOperation);
+            validator.ValidateTag("Result", x =>
+            {
+                if (expectedBoolConstraint is bool expected)
+                {
+                    x.Should().NotBeNull($"we expect {expectedBoolConstraint} on the result");
+                    x.HasConstraint(BoolConstraint.From(expected)).Should().BeTrue("we should have learned that result is {0}", expected);
+                }
+                else
+                {
+                    x.Should().BeNull("we should not learn about the state of result");
+                }
+            });
+        }
+
+        private static void ValidateSetBoolConstraint_TwoStates(string testedSymbolName, string isPattern, OperationKind expectedOperation, string expectedForTrue, string expectedForFalse)
+        {
+            var validator = CreateSetBoolConstraintValidator($"{testedSymbolName} {isPattern}");
+            validator.ValidateContainsOperation(expectedOperation);
+            validator.TagValues("Result").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(BoolConstraint.True))
+                .And.ContainSingle(x => x.HasConstraint(BoolConstraint.False));
+            AssertSymbol(BoolConstraint.True, expectedForTrue);
+            AssertSymbol(BoolConstraint.False, expectedForFalse);
+
+            void AssertSymbol(BoolConstraint branchConstraint, string expected)
+            {
+                var result = validator.Symbol("result");
+                var testedSymbol = validator.Symbol(testedSymbolName);
+                var state = validator.TagStates("Result").Single(x => x[result].HasConstraint(branchConstraint));
+                if (expected is null)
+                {
+                    state[testedSymbol].Should().BeNull("we should not learn about the tested symbol in {0} branch", branchConstraint);
+                }
+                else
+                {
+                    var testedSymbolConstraints = state[testedSymbol].AllConstraints.Select(x => x.ToString()).OrderBy(x => x).JoinStr(", "); // Rename for better assertion message
+                    testedSymbolConstraints.Should().Be(expected, "we are in {0} branch", branchConstraint);
+                }
```

Comment:
This should give better assertion messages if things go wrong. I also inlined the duplicate `state[testedSymbol]`.
```suggestion
                var testedSymbolicValue = validator.TagStates("Result").Should().ContainSingle(x => x[result].HasConstraint(branchConstraint)).Which[testedSymbol];
                if (expected is null)
                {
                    testedSymbolicValue.Should().BeNull("we should not learn about the tested symbol in {0} branch", branchConstraint);
                }
                else
                {
                    var testedSymbolConstraints = testedSymbolicValue.AllConstraints.Select(x => x.ToString()).OrderBy(x => x).JoinStr(", "); // Rename for better assertion message
                    testedSymbolConstraints.Should().Be(expected, "we are in {0} branch", branchConstraint);
                }
```

Reply:
Will do in the next PR